src/ray/raylet/node_manager.cc

```diff
@@ -806,34 +806,49 @@ void NodeManager::ProcessDisconnectClientMessage(
 
   // If the client has any blocked tasks, mark them as unblocked. In
   // particular, we are no longer waiting for their dependencies.
-  if (worker) {
-    while (!worker->GetBlockedTaskIds().empty()) {
-      // NOTE(swang): HandleTaskUnblocked will modify the worker, so it is
-      // not safe to pass in the iterator directly.
-      const TaskID task_id = *worker->GetBlockedTaskIds().begin();
-      HandleTaskUnblocked(client, task_id);
+  if (is_worker && worker->IsDead()) {
+    // Don't need to unblock the client if it's a worker and is already dead.
+    // Because in this case, its task is already cleaned up.
+    RAY_LOG(DEBUG) << "Skip unblocking worker because it's already dead.";
+  } else {
+    if (worker) {
+      while (!worker->GetBlockedTaskIds().empty()) {
+        // NOTE(swang): HandleTaskUnblocked will modify the worker, so it is
+        // not safe to pass in the iterator directly.
+        const TaskID task_id = *worker->GetBlockedTaskIds().begin();
+        HandleTaskUnblocked(client, task_id);
+      }
     }
   }
 
-  // Remove the dead client from the pool and stop listening for messages.
   if (is_worker) {
-    // The client is a worker. Handle the case where the worker is killed
-    // while executing a task. Clean up the assigned task's resources, push
-    // an error to the driver.
-    // (See design_docs/task_states.rst for the state transition diagram.)
-    const TaskID &task_id = worker->GetAssignedTaskId();
-    if (!task_id.is_nil() && !worker->IsDead()) {
-      // If the worker was killed intentionally, e.g., when the driver that created
-      // the task that this worker is currently executing exits, the task for this
-      // worker has already been removed from queue, so the following are skipped.
-      const Task &task = local_queues_.RemoveTask(task_id);
-      // Handle the task failure in order to raise an exception in the
-      // application.
-      TreatTaskAsFailed(task);
+    // The client is a worker.
+    if (worker->IsDead()) {
+      // If the worker was killed by us because the driver exists,
+      // treat it as intentionally disconnected.
+      intentional_disconnect = true;
+    }
+
+    const ActorID &actor_id = worker->GetActorId();
+    if (!actor_id.is_nil()) {
+      // If the worker was an actor, update actor state, reconstrut the actor if needed,
+      // and clean up actor's tasks if the actor is permanently dead.
+      HandleDisconnectedActor(actor_id, true, intentional_disconnect);
+    }
 
-      const JobID &job_id = worker->GetAssignedDriverId();
+    const TaskID &task_id = worker->GetAssignedTaskId();
+    if (!task_id.is_nil()) {
```

Comment:
```suggestion
    if (!task_id.is_nil() && !worker->IsDead()) {
```

Reply:
It looks like all the code in this `if block` requires the worker is not dead, it would look simpler to extract the IsDead check to the if check itself, thus we don't need to have it multiple times in the block.

Reply:
Agreed, it seems like the nested cases should be something like this:
```c++
if (is_worker) {
  if (worker->IsDead()) {
  } else {
  }
} else if (is_driver) {
}
```

Reply:
@stephanie-wang I don't really understand your suggestion.

Reply:
I just found the if-checks a little hard to read in this section, although I think it is mainly because of the existing code. Looking at the code more closely, though, I guess my suggestion isn't quite right...